### PR TITLE
Remove contraction from region inference

### LINKED ISSUE #29048: Nightly regression: Incorrect "parameter type may not live long enough" warning
Testcase:

```rust
use std::marker::PhantomData;
pub struct WebDriverHttpApi<U> {
    _routes: PhantomData<U>,
}
impl <U> WebDriverHttpApi<U> {
    pub fn new(extension_routes:Vec<(&str, U)>) -> WebDriverHttpApi<U> {
        for &(ref _url, ref _extension_route) in extension_routes.iter() {}
        unimplemented!()
    }
}
fn main(){}
```

```
<anon>:7:67: 7:73 warning: the parameter type `U` may not live long enough [E0311]
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
                                                                           ^~~~~~
<anon>:7:67: 7:73 help: consider adding an explicit lifetime bound for `U`
<anon>:6:72: 9:6 note: the parameter type `U` must be valid for the anonymous lifetime #1 defined on the block at 6:71...
<anon>:6     pub fn new(extension_routes:Vec<(&str, U)>) -> WebDriverHttpApi<U> {
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
<anon>:8         unimplemented!()
<anon>:9     }
<anon>:7:67: 7:73 note: this warning results from recent bug fixes and clarifications; it will become a HARD ERROR in the next release. See RFC 1214 for details.
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
                                                                           ^~~~~~
<anon>:7:67: 7:73 note: ...so that the reference type `&collections::vec::Vec<(&str, U)>` does not outlive the data it points at
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
                                                                           ^~~~~~
<anon>:7:67: 7:73 warning: the parameter type `U` may not live long enough [E0311]
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
                                                                           ^~~~~~
<anon>:7:67: 7:73 help: consider adding an explicit lifetime bound for `U`
<anon>:6:72: 9:6 note: the parameter type `U` must be valid for the anonymous lifetime #1 defined on the block at 6:71...
<anon>:6     pub fn new(extension_routes:Vec<(&str, U)>) -> WebDriverHttpApi<U> {
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
<anon>:8         unimplemented!()
<anon>:9     }
<anon>:7:67: 7:73 note: this warning results from recent bug fixes and clarifications; it will become a HARD ERROR in the next release. See RFC 1214 for details.
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
                                                                           ^~~~~~
<anon>:7:67: 7:73 note: ...so that the reference type `&[(&str, U)]` does not outlive the data it points at
<anon>:7         for &(ref _url, ref _extension_route) in extension_routes.iter() {}
                                                                           ^~~~~~
```

This gives no warnings on beta or stable.

### PR DESCRIPTION
**Pull request by nikomatsakis**
This fixes #29048 (though I think adding better transactional support would be a better fix for that issue, but that is more difficult). It also simplifies region inference and changes the model to a pure data flow one, as discussed in [this internals thread](https://internals.rust-lang.org/t/rough-thoughts-on-the-impl-of-region-inference-mir-etc/2800). I am not 100% sure though if this PR is the right thing to do -- or at least maybe not at this moment, so thoughts on that would be appreciated.

r? @pnkfelix 
cc @arielb1

### CODE REVIEW

File: src/librustc/middle/def_id.rs [hunk 1]
```diff
@@ -61,12 +61,16 @@ impl fmt::Debug for DefId {
         // Unfortunately, there seems to be no way to attempt to print
         // a path for a def-id, so I'll just make a best effort for now
         // and otherwise fallback to just printing the crate/node pair
-        try!(ty::tls::with_opt(|opt_tcx| {
-            if let Some(tcx) = opt_tcx {
-                try!(write!(f, " => {}", tcx.item_path_str(*self)));
-            }
-            Ok(())
-        }));
+        if self.is_local() { // (1)
```
Comment (arielb1) [inline src/librustc/middle/def_id.rs]: Does this fix `RUST_LOG=rustc::middle::traits`?

**Comment by nikomatsakis** [comment]
Interestingly, this PR fixes a bug that @dherman was encountered in his nanny project. I need to investigate that a bit more deeply though.

File: src/librustc/middle/def_id.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc/middle/def_id.rs]: @arielb1 

> Does this fix RUST_LOG=rustc::middle::traits?

Yes, I expect so. I think I was having problems with `RUST_LOG=rustc_typeck`, but I suspect it's the same issue. And yes, a fallible version of `item_path_str` would be better.

**Comment by nikomatsakis** [comment]
I looked into @dherman's case. It comes down to a trait method defined like this:

```rust
fn foo<'root, T:Trait<'root>>(t: T) -> &i32
```

the point being that `'root` does not appear in the signature in anyway, and hence has no incoming constraints. In some sense this PR is a reasonable way to fix that problem, though I think that the fix I *really* want is to guarantee that all type/lifetime parameters to a fn outlive calls to that fn. But that requires some changes to the language to achieve completely. Still, I am somewhat surprised that there aren't incoming edges to the variable from the point where the parameters are instantiated -- I think a reasonable change might be to make it so that when you create a region variable, you must provide a lower bound (which could be `ty::ReEmpty`). I've started a PR to that effect a few times but never quite finished it.

**Comment by nikomatsakis** [comment]
@arielb1 

>  Isn't SupSupConflict dead now? You may want to remove it.

yes, done.

**Comment by arielb1** [comment]
@nikomatsakis 

I am unable to reproduce @dherman's case, but its not like putting a restriction on `'a` would be helpful in any way, because of the "lifetime truncation" behaviour.

**Comment by nikomatsakis** [comment]
@arielb1 what do you mean by "lifetime truncation"?

**Comment by arielb1** [comment]
@nikomatsakis 

The fact that you can substitute an fn with some lifetime parameters that don't appear in the signature and then call it after these lifetime parameters are dead.

A reasonable argument for the soundness of this is that, because lifetime bounds in Rust are never strict, if you LUB all the lifetimes in a validate predicate with some lifetime you still get something valid - this is basically the reason lifetime erasure works (and therefore I call it "truncation"). 

For example, in your code:
```Rust
trait Trait<'root> {}
impl<'a,'root> Trait<'root> for &'a i32 where 'a: 'root {}
fn foo<'root, T:Trait<'root>>(t: T) -> i32 {
    /* ... - your original signature doesn't specify the return type's lifetime */
}

fn main() {
    'v: {
        let v = &2;
        ('t: { foo::<'t, &'v i32> })(&v);
    }
}
```

`foo` is substituted with the lifetime parameter `'t`, which is valid as `&'v i32: Trait<'t>` holds (because of `'v: 't`). On the other hand, `'t` is not live when we call `foo`, so for soundness we can LUB `'t` with the lifetime of the call to get some lifetime `'t2`, LUB `'v` with the lifetime of the call to get `'v` back, and we have that `foo::<'v, &'v i32>` is a valid substitution.

This, of course, breaks horribly if we have types that can't be bought fresh to some lifetime - e.g. unloadable code, but we already know that.

**Comment by nikomatsakis** [comment]
OK, yes. This is what I was saying I would potentially like to fix, though
it requires language changes.

**Comment by nikomatsakis** [comment]
@pnkfelix would you be able to review this sooner rather than later? It does fix a regression, which will be branched into beta any day now.

**Comment by pnkfelix** [comment]
okay it took me a little bit longer than I expected to wrap my head around this.  You left in the `fn contraction` method, so at first I thought something more subtle was going on, but all it now does is check that the `ConstrainVarSubReg` constraints hold.

It seems like the code is just calling out for a simplification pass; e.g. the factoring of `fn check_node` within the body of `fn contract_node`, which made some sense when there was a separate `fn adjust_node` path, just seems like unnecessary abstraction.

In addition, it would be nice if the README were updated, removing the [discussion of processing the Contracting nodes](https://github.com/nikomatsakis/rust/tree/0f192f1df4a1cd552ef58bdef816f9bb624cac4d/src/librustc/middle/infer/region_inference#our-current-algorithm).

But there's no reason to hold up this PR for those kinds of changes.

**Comment by pnkfelix** [comment]
@nikomatsakis r=me once you've addressed the travis failure in some reasonable manner.

**Comment by nikomatsakis** [comment]
@pnkfelix 

> It seems like the code is just calling out for a simplification pass; e.g. the factoring of fn check_node within the body of fn contract_node, which made some sense when there was a separate fn adjust_node path, just seems like unnecessary abstraction.

Yes, I agree, I was just being hasty. I will try to clean it up a bit.

> In addition, it would be nice if the README were updated, removing the discussion of processing the Contracting nodes.

agreed

**Comment by arielb1** [comment]
`region_inference/README.md` needs some rework to account for unboxed closures and the new higher-rank inference system (I think).

**Comment by SimonSapin** [comment]
Does the beta-nominated tag means this PR is to be backported to 1.5?

I’d appreciate this: at the moment I have code that compiles on stable and nightly, but not 1.5.0-beta.2: https://travis-ci.org/servo/rust-cssparser/builds/88377023

**Comment by alexcrichton** [comment]
@SimonSapin yeah if accepted this'll be backported to 1.5

**Comment by brson** [comment]
@rust-lang/lang @SimonSapin wants this backported soon. Can you review?

**Comment by nikomatsakis** [comment]
@brson I think this is really a @rust-lang/compiler issue. But I'm inclined to accept it this for beta backport.

**Comment by pnkfelix** [comment]
+1 to backport

**Comment by brson** [comment]
[Backport PR](https://github.com/rust-lang/rust/pull/29592) is up when you are ready.

**Comment by nikomatsakis** [comment]
Accepting for beta.

**Pull request merged**